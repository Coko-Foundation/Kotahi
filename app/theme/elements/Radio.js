import { css, keyframes } from 'styled-components'
import { th } from '@pubsweet/ui-toolkit'

const checking = keyframes`
  0% {
    transform: scale(0.8);
  }
  20% {
    transform: scale(1.2);
  }
  80% {
    transform: scale(1);
  }
  100% {
    transform: scale(1);
  }
`
export default {
  Root: css`
    transition: all ${th('transitionDuration')};

    &:hover {
      span {
        color: ${props =>
          props.checked ? 'inherit' : props.theme.colorPrimary};

        &:before {
          animation-name: ${props => (props.checked ? 'none' : checking)};
          animation-duration: ${th('transitionDuration')};
          box-shadow: 0 0 0 ${th('borderWidth')}
            ${props =>
              props.checked ? 'currentColor' : props.theme.colorPrimary};
        }
      }
    }
  `,
  Label: css`
    font-style: italic;

    &:before {
      content: ' ';
      display: inline-block;
      vertical-align: middle;
      width: calc(${th('gridUnit')} * 2);
      height: calc(${th('gridUnit')} * 2);
      margin-left: ${th('gridUnit')};
      margin-right: ${th('gridUnit')};

      /* This is not a real border (box-shadow provides that), so not themed as such */
      border: calc(${th('gridUnit')} / 4) solid white;
      border-radius: 50%;

      transition: border ${th('transitionDuration')}
        ${th('transitionTimingFunction')};
      color: ${props => (props.color ? props.color : props.theme.colorText)};
      background: ${props => (props.checked ? 'currentColor' : 'transparent')};
      box-shadow: 0 0 0 ${th('borderWidth')} currentColor;
    }
  `,
  Input: css`
    position: absolute;
    opacity: 0;
    z-index: -1;

    &:focus + span:before {
      box-shadow: 0 0 ${th('borderWidth')} calc(${th('borderWidth')} * 2)
        ${th('colorPrimary')};
    }
  `,
}
