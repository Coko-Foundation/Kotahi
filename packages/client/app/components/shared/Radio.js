import React from 'react'
import styled, { css } from 'styled-components'
import { th, override } from '@pubsweet/ui-toolkit'

const Label = styled.span`
  ${props =>
    props.disabled &&
    css`
      cursor: not-allowed;
    `};

  ${override('ui.Label')};
  ${override('ui.Radio.Label')};
`

const Input = styled.input`
  margin-right: ${th('gridUnit')};
  ${override('ui.Radio.Input')};
`

const Root = styled.label`
  align-items: center;
  color: ${props => (props.color ? props.color : props.theme.colorText)};
  cursor: pointer;
  display: ${props => (props.inline ? 'inline-flex' : 'flex')};
  min-height: calc(${th('gridUnit')} * 3);

  ${override('ui.Radio')};
`

// eslint-disable-next-line import/prefer-default-export
export const Radio = ({
  className,
  disabled,
  color,
  inline,
  name,
  value,
  label,
  checked,
  required,
  onChange,
}) => (
  <Root checked={checked} className={className} color={color} inline={inline}>
    <Input
      checked={checked}
      disabled={disabled}
      name={name}
      onChange={onChange}
      required={required}
      type="radio"
      value={value}
    />
    <Label checked={checked} color={color} disabled={disabled}>
      {label}
    </Label>
  </Root>
)
