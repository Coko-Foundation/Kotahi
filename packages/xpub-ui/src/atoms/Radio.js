import React from 'react'
import classes from './Radio.local.css'

const Radio = ({ name, value, label, checked, required, handleChange }) => (
  <label className={classes.root}>
    <input
      className={classes.input}
      type="radio"
      name={name}
      value={value}
      checked={checked}
      required={required}
      onChange={event => handleChange(event.target.value)}/>
    {label}
  </label>
)

export default Radio
