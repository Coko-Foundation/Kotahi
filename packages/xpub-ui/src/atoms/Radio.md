A radio button.

```js
<Radio 
  name="foo" 
  handleChange={event => console.log(event)}/>
```

A checked radio button.

```js
<Radio 
  name="bar" 
  checked 
  handleChange={event => console.log(event)}/>
```

A radio button with a label.

```js
<Radio 
  name="foo" 
  label="Foo"
  handleChange={event => console.log(event)}/>
```
