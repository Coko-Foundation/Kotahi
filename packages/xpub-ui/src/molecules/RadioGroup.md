A group of radio buttons.

```js
initialState = { foo: undefined };

const options = [
  {
    value: 'one',
    label: 'One'
  },
  {
    value: 'two',
    label: 'Two'
  },
  {
    value: 'three',
    label: 'Three'
  }
];

<RadioGroup 
  options={options} 
  name="foo"
  value={state.foo} 
  handleChange={foo => setState({ foo })}/>
```
