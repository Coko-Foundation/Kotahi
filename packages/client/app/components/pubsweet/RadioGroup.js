/* eslint-disable react/destructuring-assignment */

import React from 'react'

import Radio from './Radio'
import Flexbox from './Flexbox'

class RadioGroup extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      value: props.value,
    }
  }

  handleChange = event => {
    const { value } = event.target
    this.setState({ value })
    this.props.onChange(value)
  }

  render() {
    const { className, disabled, inline, name, options, required } = this.props
    const { value } = this.state

    return (
      <Flexbox column={!inline}>
        {options.map(option => (
          <Radio
            checked={option.value === value}
            className={className}
            color={option.color}
            disabled={disabled || option.disabled}
            inline={inline}
            key={option.value}
            label={option.label}
            name={name}
            onChange={this.handleChange}
            required={required}
            value={option.value}
          />
        ))}
      </Flexbox>
    )
  }
}

export default RadioGroup
