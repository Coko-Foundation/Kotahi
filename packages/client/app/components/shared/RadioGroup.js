import React from 'react'
import styled from 'styled-components'
import { Radio } from './Radio'

const Flexbox = styled.div`
  display: flex;
  flex-direction: ${props => (props.column ? 'column' : 'row')};
  justify-content: ${props => (props.center ? 'center' : 'left')};
`

// eslint-disable-next-line import/prefer-default-export
export const RadioGroup = ({
  options,
  value: radioValue,
  'data-testid': dataTestid,
  className,
  disabled,
  inline,
  required,
  onChange,
}) => {
  const handleChange = event => {
    const { value } = event.target
    onChange(value)
  }

  return (
    <Flexbox column={!inline} data-testid={dataTestid}>
      {options.map(option => (
        <Radio
          checked={option.value === radioValue}
          className={className}
          color={option.color}
          disabled={disabled || option.disabled}
          inline={inline}
          key={option.value}
          label={option.label}
          name={option.name}
          onChange={handleChange}
          required={required}
          value={option.value}
        />
      ))}
    </Flexbox>
  )
}
