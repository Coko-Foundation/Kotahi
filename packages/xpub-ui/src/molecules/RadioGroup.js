import React from 'react'
import Radio from '../atoms/Radio'

class RadioGroup extends React.Component {
  constructor (props) {
    super(props)

    this.state = {
      value: props.value
    }
  }

  handleChange = event => {
    this.setState({
      value: event.target.value
    })

    this.props.onChange(event)
  }

  render () {
    const { name, options, required } = this.props
    const { value } = this.state

    return (
      <div>
        {options.map(option => (
          <Radio
            key={option.value}
            name={name}
            required={required}
            value={option.value}
            label={option.label}
            checked={option.value === value}
            onChange={this.handleChange}/>
        ))}
      </div>
    )
  }
}

export default RadioGroup
