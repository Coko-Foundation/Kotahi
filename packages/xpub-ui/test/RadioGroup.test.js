import React from 'react'
import { clone } from 'lodash'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'

import Radio from '../src/atoms/Radio'
import RadioGroup from '../src/molecules/RadioGroup'

const props = {
  name: 'TestName',
  options: [
    {
      label: 'Yes',
      value: 'yes',
    },
    {
      label: 'No',
      value: 'no',
    },
    {
      label: 'Maybe',
      value: 'maybe',
    },
  ],
  required: true,
  value: undefined,
}

const wrapper = shallow(<RadioGroup {...props} />)
const radios = wrapper.find(Radio)

describe('Radio Group', () => {
  test('Snapshot', () => {
    const tree = renderer.create(<RadioGroup {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Renders a number of radio buttons', () => {
    expect(wrapper.is('div')).toBeTruthy()

    const len = props.options.length
    expect(wrapper.children()).toHaveLength(len)

    let i = 0

    while (i < len) {
      const child = wrapper.childAt(i)
      expect(child.is(Radio)).toBeTruthy()

      i++
    }

    expect(radios).toHaveLength(len)
  })

  test('Radios get the correct props', () => {
    const radioComps = radios.getElements()
    let i = 0

    while (i < props.options.length) {
      const radio = radioComps[i]
      const radioProps = radio.props

      expect(radioProps.label).toEqual(props.options[i].label)
      expect(radioProps.value).toEqual(props.options[i].value)
      expect(radioProps.name).toEqual(props.name)
      expect(radioProps.required).toEqual(props.required)

      i++
    }
  })

  test('Value should match the checked radio button', () => {
    // With no radio button selected
    const radioComps = radios.getElements()
    let i = 0

    while (i < props.options.length) {
      const radio = radioComps[i]
      const radioProps = radio.props

      expect(radioProps.checked).toBeFalsy()
      i++
    }

    // With the first radio button selected
    // (re-initialise the wrapper with changed props)
    const newProps = clone(props)
    newProps.value = 'yes'

    const newWrapper = shallow(<RadioGroup {...newProps} />)
    const newRadios = newWrapper.find(Radio)
    const newRadioComps = newRadios.getElements()

    i = 0

    while (i < props.options.length) {
      const radio = newRadioComps[i]
      const radioProps = radio.props

      if (i === 0) {
        expect(radioProps.checked).toBeTruthy()
      } else {
        expect(radioProps.checked).toBeFalsy()
      }

      i++
    }
  })
})
