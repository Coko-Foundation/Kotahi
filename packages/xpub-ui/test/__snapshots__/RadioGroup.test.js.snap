// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio Group Snapshot 1`] = `
<div>
  <label
    className="root"
    style={
      Object {
        "color": "black",
      }
    }
  >
    <input
      checked={false}
      className="input"
      name="TestName"
      onChange={[Function]}
      required={true}
      type="radio"
      value="yes"
    />
    <span
      className="pseudoInput"
      style={
        Object {
          "background": "transparent",
        }
      }
    >
       
    </span>
    <span
      className="label"
    >
      Yes
    </span>
  </label>
  <label
    className="root"
    style={
      Object {
        "color": "black",
      }
    }
  >
    <input
      checked={false}
      className="input"
      name="TestName"
      onChange={[Function]}
      required={true}
      type="radio"
      value="no"
    />
    <span
      className="pseudoInput"
      style={
        Object {
          "background": "transparent",
        }
      }
    >
       
    </span>
    <span
      className="label"
    >
      No
    </span>
  </label>
  <label
    className="root"
    style={
      Object {
        "color": "black",
      }
    }
  >
    <input
      checked={false}
      className="input"
      name="TestName"
      onChange={[Function]}
      required={true}
      type="radio"
      value="maybe"
    />
    <span
      className="pseudoInput"
      style={
        Object {
          "background": "transparent",
        }
      }
    >
       
    </span>
    <span
      className="label"
    >
      Maybe
    </span>
  </label>
</div>
`;
